src/main/java/org/owasp/esapi/reference/DefaultEncoder.java

```diff
@@ -343,7 +344,18 @@ public String encodeForLDAP(String input, boolean encodeWildcards) {
                     sb.append("\\00");
                     break;
                 default:
-                    sb.append(c);
+                    if (c >= 0x80) {
+                        try {
+                            final byte[] u = String.valueOf(c).getBytes("UTF-8");
```

Comment:
Something feels a bit off here.  char is 16bit.... but the data in `input` could be a string of characters that won't live through this translation.  I'm thinking of my unit tests over here:

```
    @Test
    public void test32BitCJK(){
        String s = "𡘾𦴩𥻂";
        String expected = "&#x2163e;&#x26d29;&#x25ec2;";
        String bad = "&#xd845;&#xde3e;&#xd85b;&#xdd29;&#xd857;&#xdec2;";
        assertEquals(false, expected.equals(bad));
        assertEquals(expected, codec.encode(new char[0], s));
    }

    @Test
    public void test32BitCJKMixedWithBmp(){
        String s = "𡘾𦴩<𥻂";
        String expected = "&#x2163e;&#x26d29;&lt;&#x25ec2;";
        String bad = "&#xd845;&#xde3e;&#xd85b;&#xdd29;&#xd857;&#xdec2;";
        assertEquals(false, expected.equals(bad));
        assertEquals(expected, codec.encode(new char[0], s));
    }
```
    
So imagine that the input is `𡘾𦴩𥻂`

I would think that there would be an overflow the second we call `char c = input.charAt(i);`

Because under the hood, java is representing `𡘾𦴩𥻂` as a UTF-16 buffer.  `𡘾` would get translated to garbage and then disaster follows.

Reply:
@xeno6696 - Can this wait until _after_ this upcoming 2.5.3.0 release? In other words, I'll phrase the question this way: Which is less wrong? The way the code presently is or the way this PR is? If you honestly think that the code in this PR is worse than what we presently have, we can exclude it from this next release, but I'm not sure I see it this way. Even if this code has the issues that you mention, I think it moves us toward the direction of the of compliance with the RFC. So, let's not let the perfect become the enemy of the good here. IMO, unless this is going to open up worse security holes in your opinion, I think the way forward should create a GitHub issue as to what needs fixed once this PR is merged and then merge this and deal with the rest of it later.

@noloader, @xeno6696 - Thoughts?

Reply:
My main worry is this:  Assuming that the rules we were breaking were breaking because (like browsers) some LDAP systems don't obey the RFC to the letter anyway, we could unintentionally break something.  That said, I'm also unsure of how many folks are using ESAPI for this anyway.  I won't stand in the way of the merge and I'm okay with punting for future issues, but then you also stated a desire to push 2.x into maintenance only which effectively bars that.  

Just putting that out there.